src/util/mod.rs

```diff
@@ -50,35 +51,20 @@ mod thread_metrics;
 
 pub const NO_LIMIT: u64 = u64::MAX;
 
-pub fn get_limit_at_size<'a, T, I>(entries: I, max: u64) -> usize
-where
-    T: Message + Clone,
-    I: IntoIterator<Item = &'a T>,
-{
-    let mut iter = entries.into_iter();
-    // If max is NO_LIMIT, we can return directly.
-    if max == NO_LIMIT {
-        return iter.count();
-    }
-
-    let mut size = match iter.next() {
-        None => return 0,
-        Some(e) => Message::compute_size(e) as u64,
-    };
-    let mut limit = 1;
-    for e in iter {
-        size += Message::compute_size(e) as u64;
-        if size > max {
-            break;
+pub fn limit_size<T: Message + Clone>(entries: &mut Vec<T>, max: u64) {
+    if max != NO_LIMIT {
+        let mut limit = 0;
+        let mut size = 0;
+        for e in entries.into_iter() {
+            size += Message::compute_size(e) as u64;
+            if size > max {
+                break;
+            }
+            limit += 1;
         }
-        limit += 1;
+        limit = cmp::max(limit, 1);
```

Comment:
Maybe check if `entries.len() <= 1` and return early is better. This method can be called frequently with only one entry in vector.